oio/zk/admin.py

```diff
@@ -0,0 +1,192 @@
+# Copyright (C) 2017-2018 OpenIO SAS, as part of OpenIO SDS
+#
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 3.0 of the License, or (at your option) any later version.
+#
+# This library is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this library.
+
+import logging
+import itertools
+import threading
+import zookeeper
+from time import time as now
+
+
+_PREFIX = '/hc'
+_PREFIX_NS = _PREFIX + '/ns'
+_acl_openbar = [{'perms': zookeeper.PERM_ALL,
+                 'scheme': 'world',
+                 'id': 'anyone'}]
+_srvtypes = (('meta0', 0, 0),
+             ('meta1', 1, 3),
+             ('meta2', 2, 2),
+             ('sqlx', 2, 2))
+
+
+def get_meta0_paths(zh, ns):
+    base = _PREFIX_NS + '/' + ns
+    path = base + '/srv/meta0'
+    yield path.strip()
+
+
+class ZkHandle(object):
+    def __init__(self, zh):
+        self._zh = zh
+
+    def get(self):
+        return self._zh
+
+    def close(self):
+        zookeeper.close(self._zh)
+        self._zh = None
+
+
+def get_connected_handles(cnxstr):
+    zookeeper.set_debug_level(zookeeper.LOG_LEVEL_INFO)
+    for shard in cnxstr.split(";"):
+        zh = zookeeper.init(shard)
+        yield ZkHandle(zh)
+
+
+def _batch_split(nodes, N):
+    """
+    Generate batches of paths with a common prefixes, and a maximal size of
+    N items.
+    """
+    last = 0
+    batch = list()
+    for x in nodes:
+        current = x[0].count('/')
+        batch.append(x)
+        if len(batch) >= N or last != current:
+            yield batch
+            batch = list()
+        last = current
+    yield batch
+
+
+def _batch_create(zh, batch):
+    sem = threading.Semaphore(0)
+    started, failed = 0, 0
+
+    def create_ignore_errors(zh, path, data):
+        def completion(*args, **kwargs):
+            global failed
```

Comment:
failed defined at l.78 is not global but has a local scope
you should use a dict, more simple (or wait python 3 for nonlocal keyword)